### PR TITLE
Design doc: Faithful `EXPLAIN <catalog item>` output

### PR DESCRIPTION
**Pull request by aalexandrov**
Design doc ([rendered](https://github.com/aalexandrov/materialize/blob/issue_20652_design/doc/developer/design/20230802_explain_running_dataflows.md)) for some catalog changes required for #20652.

I am trying to get the approval of @MaterializeInc/surfaces before I start working on this.

### Motivation

  * This PR adds a known-desirable feature.

Outlines an implementation plan for #20652.

### Tips for reviewer

N/A

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (jkosh44) [inline doc/developer/design/20230802_explain_running_dataflows.md:75]: FWIW, this is the full struct that is stored on disk for catalog items: https://github.com/MaterializeInc/materialize/blob/a0be13cb9dd52c6e60faa500b74d631bdbdcc515/src/adapter/src/catalog/storage.rs#L2049-L2061

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (jkosh44) [inline doc/developer/design/20230802_explain_running_dataflows.md:86]: Is this a collection of `GlobalId`s?

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (jkosh44) [inline doc/developer/design/20230802_explain_running_dataflows.md:118]: What does the `UsedIndexes` struct look like? Also is that information derivable from the optimized or physical plan?

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:86]: Yes, at the moment it is:

```rust
pub struct UsedIndexes(Vec<GlobalId>);
```

In the future this might change to disambiguate indexes used for full scans vs point lookups, but if we have this collection we can re-construct the rest lazily.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:75]: Thanks. I've also referenced this in "Appendix B".

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [inline doc/developer/design/20230802_explain_running_dataflows.md:240]: @bkirwi: what is the current way to encode 

> the current persist MFP behavior (enabled or not)

based on the current set of feature flag values?

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (teskje) [reply doc/developer/design/20230802_explain_running_dataflows.md:86]: Sorry for the unrelated question, but why do we need this separate list of indexes when we have a plan? I suppose that means the plan doesn't specify `Get(index_id)` directly but references `used_indexes` somehow instead?

Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:86]: >  I suppose that means the plan doesn't specify `Get(index_id)` directly but references used_indexes somehow instead?

Exactly. I wish we had `IdxFetch(index_id)` and `IdxLookup(index_id, <expr>)` nodes, but we don't. We need to do a a pass at the plan contextualized with an (immutable for the duration of the pass) set of indexes in order to decide which indexes are actually going to be picked up at rendering time.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (teskje) [inline doc/developer/design/20230802_explain_running_dataflows.md:176]: I think you want to delete this!

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (jkosh44) [reply doc/developer/design/20230802_explain_running_dataflows.md:118]: What happens if one of the indexes in `UsedIndexes` is deleted? Then `UsedIndexes` will contain a dangling reference. Or is that ok because we never actually try and look the index up?

**Review by teskje — APPROVED** [review]
I like this!

Two things we should keep in mind:

* This will interact with @antiguru's current work on `ALTER MATERIALIZED VIEW ... SET CLUSTER` in that moving an MV to a different cluster requires a re-optimization. We have to make sure to not forget updating the new bookkeeping fields proposed here.
* If we implement this before getting to `EXPLAIN CREATE ...`, do view take away from users the ability to check their new improved view definitions before deploying them? For example, I can imagine a scenario where a user installs an MV, then realizes that it would have benefitted from an index, then installs the missing index, then invokes `EXPLAIN MATERIALIZED VIEW` to check whether the index is picked up as expected, and then re-creates the MV to actually pick up the index. This workflow stops to work if this design is implemented. That's maybe fine because nobody attempts this workflow, idk.
* More importantly: If this is implemented before `EXPLAIN CREATE ...` do we still have a syntax to show how an MV will look like before it is created? I believe `EXPLAIN SELECT ...` doesn't work because one-off SELECTs get a different monotonicity analysis than long-running dataflows.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:118]: I need to dig the place of the code where this is handled, but my understanding is that:

1. We will delete the catalog entry.
2. The dataflow that maintains the index will continue do exist as long as it has dependencies.
3. When we restart `environmentd`, we will re-optimize and re-isntall the materialized view, but at that point the deleted `GlobalId` will not land in the updated `used_indexes`.

Comment (ggevay) [reply doc/developer/design/20230802_explain_running_dataflows.md:118]: Btw. this is the issue about the problem that currently you are allowed to DROP an index that is in use, resulting in the weird behavior that Alex explained above: https://github.com/MaterializeInc/materialize/issues/20334

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (bkirwi) [reply doc/developer/design/20230802_explain_running_dataflows.md:240]: Stepping back a bit, the filter pushdown stuff has two parts:
1. The static-analysis bit: given a particular query, which of these filters can we analyze well enough to push down a filter for. This is a pure function of the query.
2. The dynamic bit: whether at runtime we fetch or ignore a particular part. This depends on lots of things: the current value of flags, the _historical_ value of flags, what the write patterns are in practice...

The current explain behaviour represents mostly 1, but also a little bit of 2 since it will behave differently depending on the live value of one flag. I think it would be fine to support this flag if it represented only 1, since it's just another way of representing data already in the plan. So I would propose to change the explain behaviour to eliminate the small dependency on 2.

I think the only thing that would need to change is to remove the check of the filtering-enabled flag at query parse time, when deciding whether/not to ignore the `filter_pushdown` explain option. If we did that, I believe your third option would be acceptable. Thoughts?

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (mjibson) [inline doc/developer/design/20230802_explain_running_dataflows.md:157]: IMO: Don't worry about doing this in multiple stages. We already have lots of other stuff that uses more memory.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (mjibson) [inline doc/developer/design/20230802_explain_running_dataflows.md:193]: We are on board. This change appears to have low risk.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (mjibson) [inline doc/developer/design/20230802_explain_running_dataflows.md:119]: What's this needed for?

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:157]: OK!

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:119]: There is some discussion about that [in Appendix A](https://github.com/aalexandrov/materialize/blob/issue_20652_design/doc/developer/design/20230802_explain_running_dataflows.md#appendix-a-explainconfig-support) and [in the associated comments](https://github.com/MaterializeInc/materialize/pull/20941#discussion_r1282259325).

Based on the comments from @bkirwi (the second link), we either (a) won't need this value, or (b) we will need to record the value of the feature flags that determine whether `filter_pushdown` was enabled when the materialized view / index was created.

**Review by mjibson — APPROVED** [review]

**Review by ggevay — APPROVED** [review]

**Comment by aalexandrov** [comment]
@mjibson re-questing re-review [of this part](https://github.com/MaterializeInc/materialize/pull/20941/files#diff-894a888e86fa417f885414089dc4910e5f46c8363d4306bd8e4fd6bc76b9a048R133-R139). We need to figure out a way to solve this problem, as at the moment I'm blocked on that.

**Comment by aalexandrov** [comment]
> This PR has higher risk. Make sure to carefully review the file hotspots. In addition to having a knowledgeable reviewer, it may be useful to add observability and/or a feature flag. [What's This?](https://help.shepherdly.io/en/articles/8251033-intro-to-shepherdly-s-risk-score)
> 
> Risk Score	Probability	Buggy File Hotspots
> 🔴 80 / 100	60%	4
> Buggy File Hotspots:

Well this is only adding a markdown file. Maybe we can teach the bot to exclude those.

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md]: At the moment its this (instead of printing the name I just print a dangling `GlobalId`):

```text
materialize=> explain materialized view mv;
                            Optimized Plan                            
----------------------------------------------------------------------
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - materialize.public.accounts_balance_idx (lookup)                +
 
(1 row)

materialize=> drop index accounts_balance_idx;
DROP INDEX
materialize=> explain materialized view mv;
                            Optimized Plan                            
----------------------------------------------------------------------
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - u9 (lookup)                                                     +
 
(1 row)
```

I can change this to print

```text
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - materialize.public.accounts_balance_idx [deleted] (lookup)      +
```

File: doc/developer/design/20230802_explain_running_dataflows.md
Comment (aalexandrov) [reply doc/developer/design/20230802_explain_running_dataflows.md:163]: I was trying to align this with the rest of the API - for example `Catalog::try_get_entry`.

**Comment by aalexandrov** [comment]
I got approval from @mjibson on the revision in #21261.

I'm about to merge this as soon as the tests pass.

**Pull request merged**
